src/sbvr-api/permissions.ts

```diff
@@ -0,0 +1,1058 @@
+																		$alias: 'k',
+																		$expr: {
+																			k: { key: apiKey }
+																		}
+																	}
+																}
+															}
+														}
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+		return getPermissions(permsFilter)
+	}, {
+		primitive: true,
+		max: env.cache.apiKeys.max,
+		maxAge: env.cache.apiKeys.maxAge,
+	}
+)
+
+export const getApiKeyPermissions = (apiKey: string, callback?: (err: Error | undefined, permissions: string[]) => void): Promise<string[]> =>
+	Promise.try(() => {
+		if (_.isString(apiKey)) {
+			return $getApiKeyPermissions(apiKey)
+		} else {
+			throw new Error('API key has to be a string, got: ' + typeof apiKey)
+		}
+	}).nodeify(callback)
+
+const getApiKeyActorId = memoize(
+	(apiKey: string) =>
+		sbvrUtils.api.Auth.get({
+			resource: 'api_key',
+			passthrough: {
+				req: rootRead
+			},
+			options: {
+				$select: 'is_of__actor',
+				$filter: {
+					key: apiKey
+				}
+			}
+		})
+		.then((apiKeys: AnyObject[]) => {
+			if (apiKeys.length === 0) {
+				throw new PermissionError()
+			}
+			const apiKeyActorID = apiKeys[0].is_of__actor.__id
+			if (apiKeyActorID == null) {
+				throw new Error('API key is not linked to a actor?!')
+			}
+			return apiKeyActorID
+		}),
+	{
+		primitive: true,
+		promise: true,
+		maxAge: env.cache.apiKeys.maxAge,
+	}
+)
+
+const checkApiKey = Promise.method((req: PermissionReq, apiKey: string) => {
+	if (apiKey == null || req.apiKey != null) {
+		return
+	}
+	return getApiKeyPermissions(apiKey)
+	.catch((err) => {
+		console.warn('Error with API key:', err)
+		// Ignore errors getting the api key and just use an empty permissions object
+		return []
+	})
+	.then((permissions) => {
+		req.apiKey = {
+			key: apiKey,
+			permissions: permissions
+		}
+	})
+})
+
+export const customAuthorizationMiddleware = (expectedScheme = 'Bearer') => {
+	if (expectedScheme == null) {
+		expectedScheme = 'Bearer';
+	}
+	expectedScheme = expectedScheme.toLowerCase()
+	return (req: express.Request, _res?: express.Response, next?: express.NextFunction): Promise<void> =>
+		Promise.try(() => {
+			const auth = req.header('Authorization')
+			if (!auth) {
+				return
+			}
+
+			const parts = auth.split(' ')
+			if (parts.length !== 2) {
+				return
+			}
+
+			const [ scheme, apiKey ] = parts
+			if (scheme.toLowerCase() !== expectedScheme) {
+				return
+			}
+
+			return checkApiKey(req, apiKey)
+		}).then(() => {
+			next && next()
+		})
+}
+
+// A default bearer middleware for convenience
+export const authorizationMiddleware = customAuthorizationMiddleware()
+
+export const customApiKeyMiddleware = (paramName = 'apikey') => {
+	if (paramName == null) {
+		paramName = 'apikey';
+	  }
+	return (req: HookReq | express.Request, _res?: express.Response, next?: express.NextFunction): Promise<void> => {
+		const apiKey = req.params[paramName] != null
```

Comment:
Why not just:

```js
const apiKey = req.params[paramName] || req.body[paramName] || req.query[paramName];
```

Reply:
The current form matches the previous behaviour which is what I was trying to stick to for the most part - ie keep everything as-is but with added types in order to minimise the chance of introducing bugs with such a large diff